### PR TITLE
Add auto approval flow for servicenow plugin

### PR DESCRIPTION
**Pull request by EdwardDowling**
Adds auto approval flow to servicenow so users with roles that have a 'schedule' annotation set will be checked to see if they are on-call in the specified rota.

PR adds the functionality to the servicenow plugin's implementation of 'app', the pr to actually use the use custom app is [here](https://github.com/gravitational/teleport.e/pull/2503) 

changelog: Add auto approval flow to servicenow plugin

### CODE REVIEW

File: integrations/access/servicenow/app.go
Comment (hugoShaka) [inline integrations/access/servicenow/app.go:316]: If I understand correctly, if we already opened an incident (or tried to open an incident) `isNew` is false and we return. This means we never reach the `postReviewNotes()` if an access request gets reviewed but is not yet approved.

I might be wrong but I think we want to update the incident with the review notes when receiving the update events. If this is really a bug, it was already present before this PR but I think we could fix it here.

File: integrations/access/servicenow/client.go
Comment (hugoShaka) [inline integrations/access/servicenow/client.go:235]: `GetOnCall()` now returns the username instead of the user email. This means the `userIsOnCall` now compares the Teleport user with the servicenow username instead of the servicenow email.

This seems like a breaking change. Was it a bug? What is the expected behaviour, checking the username or the email? If we want to do this breaking change we might want to add something in the release notes.

File: integrations/access/servicenow/app.go
Comment (marcoandredinis) [inline integrations/access/servicenow/app.go:276]: Should we include the `reqID` as a log context field?

```go
log = log.WithField("reqId", reqID)
```

File: integrations/access/servicenow/app.go
Comment (marcoandredinis) [reply integrations/access/servicenow/app.go:316]: We are not checking for auto-approval if there's already a review :thinking: 
It should be ok, but I was expecting to see the auto-approval logic to be applied even if there's an approval already

Eg it requires two approvals, but only one was given, and now the auto-approval can't kick in

File: integrations/access/servicenow/client.go
Comment (EdwardDowling) [reply integrations/access/servicenow/client.go:235]: This function was not actually called prior to this PR and some changes in [2503](https://github.com/gravitational/teleport.e/pull/2503)

File: integrations/access/servicenow/app.go
Comment (EdwardDowling) [reply integrations/access/servicenow/app.go]: What would you suggest as the ideal way the plugin should handle this?
Ill update it so the postReviewNotes gets called even if the PR is not approved yet.
But for the case of access requests that require more than 1 approval should the plugin only look at the request once it only needs 1 more approval to be approved?

Comment (marcoandredinis) [reply integrations/access/servicenow/app.go]: Every review should be logged as a note in ServiceNow
When it tries to auto-approve, it shouldn't matter whether it has 0 reviews or N reviews, it should be approved regardless.

Does the above look correct?
Sorry, I might be overlooking something

File: integrations/access/servicenow/app.go
Comment (marcoandredinis) [inline integrations/access/servicenow/app.go:272]: You can remove the `reqID` because it's already part of the logger

File: integrations/access/servicenow/app.go
Comment (marcoandredinis) [inline integrations/access/servicenow/app.go:280]: Just to ensure I'm following the logic here
If we have at least one review, we don't try to auto-approve. Is that it?

File: integrations/access/servicenow/config.go
Comment (marcoandredinis) [inline integrations/access/servicenow/config.go:37]:
```suggestion
	// the Teleport auth server. The ServiceNow app will create a gRPC-based
```

File: integrations/access/servicenow/app.go
Comment (EdwardDowling) [reply integrations/access/servicenow/app.go:280]: That section is just to post the review notes if there are some already by the time the plugin gets to that access request.

File: integrations/access/servicenow/app.go
Comment (EdwardDowling) [reply integrations/access/servicenow/app.go]: Currently it just approves it in the form of an access review but that will still only count as one of the required reviews. and it will do this no matter how many reviews there are already. Is this what you were thinking or should it be the case that it should fully approve the request instead of just counting as a single approval?

File: integrations/access/servicenow/app.go
Comment (marcoandredinis) [reply integrations/access/servicenow/app.go:277]: You are right, I'm sorry

File: integrations/access/servicenow/app.go
Comment (marcoandredinis) [reply integrations/access/servicenow/app.go]: I thought that our auto-approval did a complete approval, and not just an access review.
That's not the case for other plugins?

Comment (EdwardDowling) [reply integrations/access/servicenow/app.go]: Looking at the pagerduty auto approval it seems to do the same thing as is done here, where the plugin just submits a single approval. https://github.com/gravitational/teleport/blob/7767bf5e0ff7704b39819e4ec144af08b7557716/integrations/access/pagerduty/app.go#L498

Comment (marcoandredinis) [reply integrations/access/servicenow/app.go]: Thank you for looking into it.
I guess my assumption was wrong, sorry.

**Review by marcoandredinis — APPROVED** [review]

**Comment by EdwardDowling** [comment]
@hugoShaka Can you take another look at this when you get a chance?

File: integrations/access/servicenow/app.go
Comment (hugoShaka) [reply integrations/access/servicenow/app.go]: If user A requests access, and it requires 3 approvals.
- The plugin creates an incident in servicenow.
- User B approves the access request. It still needs 2 approvals to be usable.
- The plugin should:
  - Update the incident message to reflect how many approvals are still required (not mandatory, not all plugins are doing this)
  - Post review notes in the incident saying User B reviewed the access request (mandatory)
  
Here, when User B approves the access request, there's already plugindata (because the AR was posted on creation) and `isNew` is false. This means the plugin will return and skip the whole "post review notes" part.

I understand we don't want to evaluate if the user is on-call for every message, only on creation, but I think the plugin must update the review notes every time. This is also what the pagerduty plugin is doing.

I suggest moving the "postReviewNotes" before the `!isNew` check.

Comment (EdwardDowling) [reply integrations/access/servicenow/app.go]: Ah i see, have moved the rpostReview notes

File: integrations/access/servicenow/app.go
Comment (hugoShaka) [inline integrations/access/servicenow/app.go:280]:
```suggestion
	if isNew {
		log.Infof("Creating servicenow incident")
		if err = a.createIncident(ctx, reqID, reqData); err != nil {
			// Even if we failed to create the incident we try to auto-approve
			return trace.NewAggregate(
				trace.WrapWithMessage(err, "creating ServiceNow incident"),
				trace.Wrap(a.tryApproveRequest(ctx, req)),
			)
		}
	}

	if reqReviews := req.GetReviews(); len(reqReviews) > 0 {
		if err = a.postReviewNotes(ctx, reqID, reqReviews); err != nil {
			return trace.Wrap(err, "posting review notes")
		}
	}

	// To minimize the count of auto-approval tries, let's only attempt it only when we have just created an incident.
	if !isNew {
		return nil
	}

	// Try to approve the request if user is currently on-call.
	return trace.Wrap(a.tryApproveRequest(ctx, req))
}
```

File: integrations/access/servicenow/app.go
Comment (hugoShaka) [inline integrations/access/servicenow/app.go:274]: I swear this is the last change 🙃 

```suggestion
			return trace.NewAggregate(
				trace.WrapWithMessage(err, "posting review notes"),
				trace.Wrap(a.tryApproveRequest(ctx, req)),
			)
```

**Review by hugoShaka — APPROVED** [review]

**Pull request merged**
